File: l4/server/server/controller.py
# -*- coding: utf-8 -*-


from twisted.internet.protocol import Protocol, ClientFactory
import sys
import json
from cStringIO import StringIO


class ControllerProtocol(Protocol):
    def __init__(self):
        self.rec = StringIO()

    def connectionMade(self):
        sys.stdout.write("Success\n")

    def connectionLost(self, reason):
        from twisted.internet import reactor
        reactor.stop()

    def dataReceived(self, data):
        self.rec.write(data)
        self.rec.flush()
        self.rec = StringIO(self.rec.getvalue())
        pickling = True
        while pickling:
            try:
                result = json.load(self.rec)
            except ValueError:
                pickling = False
            except EOFError:
                self.rec.close()
                self.rec = StringIO()
                pickling = False
            except Exception as e:
                self.factory.fail(e)
                pickling = False
            else:
                keys = result.keys()
                if keys == ['welcome']:
                    print(result['welcome'])
                    self.send()
                elif keys == ['state']:
                    self.write(result['state'])
                    self.transport.loseConnection()
                self.rec = StringIO()

    def send(self):
        self.transport.write(json.dumps({'state': None}))

    def write(self, data):
        sys.stdout.write("Server state:\n{}".format(
            "".join(["{}: {}\n".format(k, v) for k, v in data.iteritems()])
        ))


class Controller(ClientFactory):
    protocol = ControllerProtocol

    def __init__(self):
        sys.stdout.write('Connecting to the server...')
        sys.stdout.flush()

    def clientConnectionFailed(self, connector, reason):
        connector.connect()


def run():
    if len(sys.argv) != 3:
        raise AttributeError(
            "Incorrect number of arguments."
            " Usage: <exec> <host> <port>"
        )
    from twisted.internet import reactor
    factory = Controller()
    reactor.connectTCP(sys.argv[1], int(sys.argv[2]), factory)
    reactor.run()

File: l4/server/controller.js
#!/usr/bin/env node


if(process.argv.length != 4) {
    console.log("Usage: <exec> <host> <port>");
    process.exit();
}
var net = require('net');
process.stdout.write("Connecting to the server...");
net.connect(process.argv[3], process.argv[2]).on('connect', function() {
    console.log("Success");
}).on('error', function(error) {
    this.setTimeout(1, function() {
        this.connect(process.argv[3], process.argv[2]);
    });
}).on('data', function(data) {
    var parsed = JSON.parse(data);
    var keys = Object.keys(parsed);
    if(keys[0] == 'welcome') {
        console.log(parsed.welcome);
        this.write(JSON.stringify({'state': null}));
    } else if(keys[0] == 'state') {
        var output = ["Server state:"];
        for(var k in parsed.state) {
            output.push(k + ": " + parsed.state[k]);
        }
        console.log(output.join('\n'));
        this.end();
    }
});
